stdlib/LinearAlgebra/src/blas.jl

```diff
@@ -964,6 +965,74 @@ Return the updated `y`.
 """
 sbmv!
 
+### spmv!, (SP) symmetric packed matrix-vector operation defined as y := alpha*A*x + beta*y.
+for (fname, elty) in ((:dspmv_, :Float64),
+                      (:sspmv_, :Float32))
+    @eval begin
+        function spmv!(uplo::AbstractChar,
+                       n::BlasInt,
+                       α::$elty,
+                       AP::Union{Ptr{$elty}, AbstractArray{$elty}},
+                       x::Union{Ptr{$elty}, AbstractArray{$elty}},
+                       incx::Integer,
+                       β::$elty,
+                       y::Union{Ptr{$elty}, AbstractArray{$elty}},
+                       incy::Integer)
+
+            ccall((@blasfunc($fname), libblas), Cvoid,
+                  (Ref{UInt8},     # uplo,
+                   Ref{BlasInt},   # n,
+                   Ref{$elty},     # α,
+                   Ptr{$elty},     # AP,
+                   Ptr{$elty},     # x,
+                   Ref{BlasInt},   # incx,
+                   Ref{$elty},     # β,
+                   Ptr{$elty},     # y, out
+                   Ref{BlasInt}),  # incy
+                  uplo,
+                  n,
+                  α,
+                  AP,
+                  x,
+                  incx,
+                  β,
+                  y,
+                  incy)
+        end
+    end
+end
+
+function spmv!(uplo::AbstractChar,
+               α::Number, AP::Union{DenseArray{T}, AbstractVector{T}}, x::Union{DenseArray{T}, AbstractVector{T}},
+               β::Number, y::Union{DenseArray{T}, AbstractVector{T}}) where {T <: BlasReal}
```

Comment:
Now that I see similar code again as in the `hpmv` PR, I started wondering why `AP`, `x` and `y` can be of type `DenseArray{T}`. For instance, in `sbmv!` the arguments `x` and `y` must be of type `AbstractVector{$elty}`. Same thing for `AP`: if the BLAS method can handle anything besides `AbstractVector`s, this should be tested then, or this signature (and `hpmv`'s narrowed).

Reply:
AbstractVector would be too broad for BLAS.

Reply:
Could be, but most if not any functions in `blas.jl` have `AbstractMatrix` and `AbstractVector` in their signatures... which may need revision. But the case here is even a bit more particular because these multiplication methods seem to require some packed storage, essentially the columns of a triangle `vec`ed. That's why I was even more confused about the rather general `DenseArray` here.

Reply:
> Now that I see similar code again as in the `hpmv` PR, I started wondering why `AP`, `x` and `y` can be of type `DenseArray{T}`. For instance, in `sbmv!` the arguments `x` and `y` must be of type `AbstractVector{$elty}`. Same thing for `AP`: if the BLAS method can handle anything besides `AbstractVector`s, this should be tested then, or this signature (and `hpmv`'s narrowed).

Thank you, that's quite an interesting point.

I would say, without claiming to know the context and reasons which might decide
whether to use `DenseArray` or not in the long run, that the approach depends on which existing
`blas.jl` function implementation one prefers to take as a reference.  It's
true that `sbmv!` does not contain `DenseArray` in its header, yet for example
`axpby!` does so and I couldn't find any obvious references to `DenseArray` in the
tests of `axpby!`.

Regarding `DenseArray` I collected some info from looking at the julia source code.
It seems to be some sort of inbuilt abstract type, used to subtype
concrete implementations with guarantees of memory contiguity beyond(?) that of
usual numeric arrays. The latter usual numeric arrays seem to be already valid
inputs to wrapped BLAS calls.

Beyond `DenseArray`'s implementation in `base` and references to it in various
function headers (for example in `blas.jl`) I found only two examples where
`DenseArray` is subtyped:

```
stdlib/SharedArrays/src/SharedArrays.jl: mutable struct SharedArray{T,N} <: DenseArray{T,N}
stdlib/Random/RNGs.jl: struct UnsafeView{T} <: DenseArray{T,1}
```

Thus I am not sure if the two options of removing `DenseArray` completely from `hpmv!`
and `spmv!` or writing extra tests are feasible, since `DenseArray` is referenced in other
BLAS function headers as well without explicit tests written and the newly added functions don't seem to
contradict this pattern. As `DenseArray` is abstract, a solution could be to create
a custom type within the tests that is a subtype of `DenseArray` but
the context and scope of doing that seems a bit misterious to me at the moment.

So any remarks, clarifications, ideas would be appreciated.

Reply:
I was reading `DenseArray` more like "`DenseMatrix`", so my concern was more that you're not enforcing the packed format. I'm not sure, but maybe you could specify the `N` parameter as `1`? Like restricting to "`DenseVector`s"? That would resolve my main concern... unless it is possible to pass a two-dimensional array for `A`, and if that was the case I think we should test the method with an array (I was referring to this situation earlier when I was suggesting to add tests).

Reply:
> I was reading `DenseArray` more like "`DenseMatrix`", so my concern was more that you're not enforcing the packed format. I'm not sure, but maybe you could specify the `N` parameter as `1`? Like restricting to "`DenseVector`s"? That would resolve my main concern... unless it is possible to pass a two-dimensional array for `A`, and if that was the case I think we should test the method with an array (I was referring to this situation earlier when I was suggesting to add tests).

Yes, your concern about not enforcing dimensionality of 1 seems valid to me, as opposed to to the cases of `dot*`, `axpy!`, `axpby!` and `nrm2` which all use `DenseArray` and have vector inputs, whereas for hpmv! and spmv! we have matrices dressed up as vectors. I am not sure though whether `DenseArray{T <: BlasReal}`  is already silently implied to be `DenseArray{T,1}` or not. The existing syntax for the other functions with truly vectorial inputs seems suggestive toward this possibility, yet then again why not change those function headers to `DenseVector` as well?

Reply:
> .... But the case here is even a bit more particular because these multiplication methods seem to require some packed storage, essentially the columns of a triangle `vec`ed....

I was considering to adopt such a conversion earlier but in my opinion requiring this would be be a compromise in speed and flexibility in some cases. I would speculate that the typical use case would be that the user is aware of BLAS and is granted to use it in a performant way. To serve the latter, there is a short function header (for example in `hpmv!`) serving as a convenience function and a longer header with a one-to-one mapping to the BLAS API if optimization is required.

Reply:
In BLAS ...

> `dot*`, `axpy!`, `axpby!` and `nrm2`

... all work for both vectors and matrices, and actually arbitrary arrays, by interpreting them as lists of numbers ("vectors"). I guess the same is true for `hpmv` and `spmv`, it's just that they are interpreted according to the packed format. So I came to realize that the dimension of the arrays is all the same to BLAS: you could provide 3-dimensional arrays, and then BLAS decides how to interpet them, as if `vec`ed, or packed, or whatever.

There is no need to provide a function that `vec`s the columns of one of the triangles. For the "full" matrix input, we have `hemv` and `symv`, even though here also only one of the triangles is read from.

Reply:
> In BLAS ...
> 
> > `dot*`, `axpy!`, `axpby!` and `nrm2`
> 
> ... all work for both vectors and matrices, and actually arbitrary arrays, by interpreting them as lists of numbers ("vectors"). I guess the same is true for `hpmv` and `spmv`, it's just that they are interpreted according to the packed format. So I came to realize that the dimension of the arrays is all the same to BLAS: you could provide 3-dimensional arrays, and then BLAS decides how to interpet them, as if `vec`ed, or packed, or whatever.
> 
> There is no need to provide a function that `vec`s the columns of one of the triangles. For the "full" matrix input, we have `hemv` and `symv`, even though here also only one of the triangles is read from.

Thank you for the final insight. I think the whole conversation was quite helpful to form a picture of what is going on.